### PR TITLE
[FLINK-23113][runtime-web] Migrate to ESLint, add Stylint and address dependency vulnerabilities

### PR DESCRIPTION
**Pull request by yangjunhan**
## What is the purpose of the change

TSLint is no longer maintained, so runtime-web should be migrating to ESLint for Angular. In addition, stylelint should be added for .less stylesheets which is originally missing.

## Brief change log

- Migrate to ESLint
- Add Stylelint
- `--fix` all `.{less,html,ts}` file according to the new rules and prettier config
- Address dependency vulnerabilities
- Enforce linting in CI

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

### CODE REVIEW

**Comment by yangjunhan** [comment]
Please note, this PR also covers _"Address vulnerabilities in Flink UI"_ [FLINK-23114](https://issues.apache.org/jira/browse/FLINK-23114).

File: flink-runtime-web/web-dashboard/src/app/interfaces/safe-any.ts
Comment (Airblader) [inline flink-runtime-web/web-dashboard/src/app/interfaces/safe-any.ts:2]: We should immediately mark this as deprecated. `SafeAny` is not safe at all, it's a hack that silences a correct warning/error. It would be better to use `unknown` instead everywhere, however that is a separate change as it isn't a drop-in replacement.

File: flink-runtime-web/web-dashboard/.eslintignore
Comment (Airblader) [inline flink-runtime-web/web-dashboard/.eslintignore:27]: We don't use yarn, do we (also below for `.yarn`)?

**Review by Airblader — CHANGES_REQUESTED** [review]
Thanks @yangjunhan for the PR! It'll be a great improvement to Flink UI to have proper linting set up. I will take a closer look next week. 

However, it would be good if we could split this into different commits, in particular actually applying all linting fixes should be its own commit. This would make the review easier, allow us to regenerate that commit upon conflicts and maybe even add it to `.git-blame-ignore-revs` (during merging). Since you mentioned the vulnerability fixes here, anything not related to tslint/eslint itself should also be a separate commit as those changes are important and useful on their own.

Another note: this currently isn't checked in the CI, correct? We need to make sure that the CI actually verifies using eslint as well (again, as a separate commit).

**Comment by yangjunhan** [comment]
> Thanks @yangjunhan for the PR! It'll be a great improvement to Flink UI to have proper linting set up. I will take a closer look next week.
> 
> However, it would be good if we could split this into different commits, in particular actually applying all linting fixes should be its own commit. This would make the review easier, allow us to regenerate that commit upon conflicts and maybe even add it to `.git-blame-ignore-revs` (during merging). Since you mentioned the vulnerability fixes here, anything not related to tslint/eslint itself should also be a separate commit as those changes are important and useful on their own.
> 
> Another note: this currently isn't checked in the CI, correct? We need to make sure that the CI actually verifies using eslint as well (again, as a separate commit).

Thank you for your advice @Airblader. Now I have split it into several individual commits and hopefully they make sense to you. Regarding the `SafeAny` issue, I already marked it as deprecated, and perhaps we will remove it in the near future.

**Comment by Airblader** [comment]
FYI, the CI currently fails because it doesn't know `npx`.

> As for the CI, can you give me some guide about the shell file that should be modified? I could not find it under ./tool/ci/.

In flink-runtime-web/pom.xml, first `npm ci` and then `npm run build` is executed. I think we should hook into the `build` script to also run the linter first. An alternative would be to create a separate npm script which does lint+build and then call that from `pom.xml`.

Once the build for this PR is passing we should intentionally introduce a commit that breaks the linter to verify that the CI is picking up on it.

**Comment by yangjunhan** [comment]
@Airblader I make some changes regarding the issues which were mentioned earlier. Let's wait for the CI to rerun and see if it passes.

**Comment by yangjunhan** [comment]
After doing some investigations and experiments on the problem `npx: Not found`, it turns out that `frontend-maven-plugin` v1.6 requires us to explicitly specify both `nodeVersion` and `npmVersion` (originally `npmVersion` was omitted, causing the not-found error). This issue was reported here [#458](https://github.com/eirslett/frontend-maven-plugin/issues/458).

Fortunately, it was solved from v1.9.0 onwards [#868](https://github.com/eirslett/frontend-maven-plugin/pull/868). Therefore, the solution would be updating the plugin version to v1.9.0+ and/or explicitly specify the `npmVersion`. Hopefully, now the CI will pass 😂.

**Comment by yangjunhan** [comment]
Hi, @Airblader. I finish enforcing the Lint during the CI build and doing several changes to make the CI to pass. There are a few things that are worthwhile to pay attention to:

- `Frontend-maven-plugin` was previously installed as v1.6 which was considered as outdated for a while. Especially there was a bug which ignoring npm installation if `npmVersion` was not explicitly stated. So I take the liberty to update this plugin to a newer version v1.9.1 and explicitly specific the npm's version in the pom.xml.
- Despite the fact of running `npm audit fix`, even with `--force`, there is still a moderate vulnerability regarding `@angular-devkit/build-angular`'s dependency, `glob-parent >=5.1.2`. The reason I leave it there is due to this issue: [#21214](https://github.com/angular/angular-cli/issues/21214). Angular CLI team is planning to solve this by upgrading to `webpack-dev-server 4` soon.

**Review by Airblader — COMMENTED** [review]
Thanks for your work on this! I think this is looking pretty good now overall. We need to wait until after 1.14 with this, I think; while it's an non-functional change, it does update things like dependencies.

File: flink-runtime-web/web-dashboard/.stylelintrc.js
Comment (yangjunhan) [reply flink-runtime-web/web-dashboard/.stylelintrc.js:43]: Well, this piece of info is taken from: https://github.com/hudochenkov/stylelint-config-hudochenkov/blob/master/order.js. The author of this config is actually a member of Stylelint team. There is a comment about this at line 22.

File: flink-runtime-web/src/main/resources/META-INF/NOTICE
Comment (yangjunhan) [reply flink-runtime-web/src/main/resources/META-INF/NOTICE:16]: Okay, thanks. In fact, I am not quite sure about this either. I will remove them if they are unnecessary.

Comment (Airblader) [reply flink-runtime-web/src/main/resources/META-INF/NOTICE:16]: Since the final release doesn't contain these dependencies they should not need to be listed. I think we can remove this.

File: flink-runtime-web/web-dashboard/.stylelintrc.js
Comment (yangjunhan) [reply flink-runtime-web/web-dashboard/.stylelintrc.js:43]: After thinking over this, maybe it is not a good idea to directly use this config's source code. I think I am going to add it as a dev dependency and extends this config in the `.stylelintrc.js` file as the README suggests.

Comment (yangjunhan) [reply flink-runtime-web/web-dashboard/.stylelintrc.js]: After thinking over it, it may not be a good idea to extract the source code from this config. Instead, I add it as a dev dependency and extend it in the `.stylelintrc.js` file as the README suggests.

File: flink-runtime-web/src/main/resources/META-INF/NOTICE
Comment (yangjunhan) [reply flink-runtime-web/src/main/resources/META-INF/NOTICE]: No problems 👌

File: flink-runtime-web/web-dashboard/package.json
Comment (Airblader) [inline flink-runtime-web/web-dashboard/package.json:10]: You're actually starting eslint as a background task and then stylelint. This will cause eslint to finish after the process has already exited, and cause a zero return code even if eslint found issues.

```suggestion
    "lint": "eslint --cache src --ext .ts,.html && stylelint \"**/*.less\"",
    "lint:fix": "eslint --fix --cache src --ext .ts,.html && stylelint \"**/*.less\" --fix",
```

**Review by Airblader — CHANGES_REQUESTED** [review]
Now that the 1.14 release branch has been cut we can proceed here and get this merged. I think overall this looks fine now, however there is one crucial issue still with the npm scripts. Afterwards this should be good to merge!

File: flink-runtime-web/web-dashboard/package.json
Comment (yangjunhan) [reply flink-runtime-web/web-dashboard/package.json]: Yep, this was a potential problem. Thanks for pointing it out.

**Review by Airblader — APPROVED** [review]
This PR looks good to me now. Thanks for working on it! I'll have to defer merging to @twalthr.

**Review by twalthr — APPROVED** [review]
Thanks for the contribution @yangjunhan and the review @Airblader. As before, I don't have much knowledge about the WebUI and need to trust you guys that the changes are valid. It would be good to have a final green build (maybe after a rebase) and all conflicts resolved. Please ping me once this is done and I will merge asap.

**Comment by Airblader** [comment]
@yangjunhan I've also noticed now that you are migrating the package-lock.json back from version 2 to 1. Can you please fix that commit to have it generated as version 2? This requires npm >= 7.

**Comment by yangjunhan** [comment]
> @yangjunhan I've also noticed now that you are migrating the package-lock.json back from version 2 to 1. Can you please fix that commit to have it generated as version 2? This requires npm >= 7.

Yep, it was caused by my local node/npm versions, and I fixed it. Please have a look to see whether this PR is ready for merging.

**Review by Airblader — APPROVED** [review]

**Comment by yangjunhan** [comment]
@twalthr I guess this PR is ready for merging now.

**Comment by MartijnVisser** [comment]
I took the liberty to also do some quick testing in the user interface. The only thing I noticed was when I opened the `Logs` or the `Stdout` tabs for both the Job Manager or Task Manager, the Chrome Devtools report multiple 404's. However, the functionality is working as expected. 
```
DevTools failed to load source map: Could not load content for http://localhost:8081/min-maps/vs/loader.js.map: HTTP error: status code 404, net::ERR_HTTP_RESPONSE_CODE_FAILURE
DevTools failed to load source map: Could not load content for http://localhost:8081/min-maps/vs/editor/editor.main.js.map: HTTP error: status code 404, net::ERR_HTTP_RESPONSE_CODE_FAILURE
DevTools failed to load source map: Could not load content for http://localhost:8081/min-maps/vs/editor/editor.main.nls.js.map: HTTP error: status code 404, net::ERR_HTTP_RESPONSE_CODE_FAILURE
DevTools failed to load source map: Could not load content for http://localhost:8081/min-maps/vs/base/worker/workerMain.js.map: Fetch through target failed: Target not supported; Fallback: HTTP error: status code 404, net::ERR_HTTP_RESPONSE_CODE_FAILURE
```

**Review by MartijnVisser — APPROVED** [review]
I see that the 404 errors also occurred in previous version (Flink 1.13.2). LGTM

**Comment by twalthr** [comment]
Thanks everyone. Merging...

**Pull request closed without merging**
